[popover] Improve focus handling

rwlbuis opened the pull request:
#### c5c4b9c561811faabfb7148569c71d58ea9f58ac
<pre>
[popover] Improve focus handling
<a href="https://bugs.webkit.org/show_bug.cgi?id=255454">https://bugs.webkit.org/show_bug.cgi?id=255454</a>

Reviewed by Ryosuke Niwa.

A lot of tests are failing because on macOS buttons are not considered to be focusable.
Start allowing this in case tabindex is set on form controls and adjust the tests accordingly.

* LayoutTests/fast/events/focus-label-legend-elements-with-tab-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-not-keyboard-focusable-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-not-keyboard-focusable.html:
* LayoutTests/platform/glib/fast/events/focus-label-legend-elements-with-tab-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/semantics/popovers/popover-not-keyboard-focusable-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-expected.txt:
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::isKeyboardFocusable const):
(WebCore::HTMLFormControlElement::isMouseFocusable const):

Canonical link: <a href="https://commits.webkit.org/263527@main">https://commits.webkit.org/263527@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0e3ac800c0d01928b8d6aeb4b3c12fc6c62d8a77

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4680 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4799 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4959 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6181 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4830 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4673 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4947 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4764 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/5059 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4755 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4841 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4190 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6191 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2330 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4181 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/9162 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4189 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4255 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5809 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4659 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3794 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4183 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/4177 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1194 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8231 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4543 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1445,7 +1445,7 @@ ExceptionOr<void> HTMLElement::hidePopoverInternal(FocusPreviousElement focusPre
 
     if (RefPtr element = popoverData()->previouslyFocusedElement()) {
-        if (focusPreviousElement == FocusPreviousElement::Yes) {
+        if (focusPreviousElement == FocusPreviousElement::Yes && contains(document().focusedElement())) {
```
This behavior change makes sense (e.g. you wouldn't want to restore focus if the person has tabbed away from the popover), but this should be in the HTML spec.

nt1m replied:
https://github.com/whatwg/html/issues/9169

nt1m replied:
Also we need to figure out if `contains()` is the correct call for popovers containing shadow DOM trees. (cc @rniwa)

rwlbuis replied:
Agreed.
Original chromium commit:
https://chromium-review.googlesource.com/c/chromium/src/+/4322607

Ahmad-S792 replied:
It might be 'containsIncludingShadowDOM'. @nt1m

rwlbuis replied:
> It might be 'containsIncludingShadowDOM'. @nt1m

Thanks, that could work, but I'll leave this for some other PR (also the relevant spec change still has not landed AFAIK).

rniwa commented on Source/WebCore/html/HTMLButtonElement.cpp:
```diff
@@ -249,4 +249,11 @@ bool HTMLButtonElement::isExplicitlySetSubmitButton() const
+bool HTMLButtonElement::isKeyboardFocusable(KeyboardEvent* event) const
+{
+    if (hasAttributeWithoutSynchronization(popovertargetAttr))
+        return isFocusable();
```
I don't think we should do this.

nt1m commented on Source/WebCore/html/HTMLButtonElement.cpp:
```diff
@@ -249,4 +249,14 @@ bool HTMLButtonElement::isExplicitlySetSubmitButton() const
+{
+    return isFocusable() && !!tabIndexSetExplicitly();
+}
+
```
I wonder if we should just change the logic at https://searchfox.org/wubkat/rev/c36758395b16531cacdf1dbdb1315a45c992ca87/Source/WebCore/html/HTMLFormControlElement.cpp#243-259 instead?

`<input type=button|submit|reset>` should behave the same as `<button>`

rwlbuis replied:
I now singled out button and <input type=button|submit|reset> to allow keyboard/mouse focus if tabindex is set, I don't think we want more form controls with the new behavior than that?

rniwa commented on Source/WebCore/html/HTMLFormControlElement.cpp:
```diff
@@ -252,6 +261,8 @@ bool HTMLFormControlElement::isMouseFocusable() const
     return HTMLElement::isMouseFocusable();
 #else
+    if (isButtonLikeControl(*this))
+        return isFocusable() && !!tabIndexSetExplicitly();
```
It seems to me that this should apply to non-button controls as well when tabIndex is explicitly set?

rwlbuis replied:
I wanted to play it safe, but now made that change, let's see what if everything shows up green.

rwlbuis committed:
[popover] Improve focus handling
https://bugs.webkit.org/show_bug.cgi?id=255454

Reviewed by Ryosuke Niwa.

A lot of tests are failing because on macOS buttons are not considered to be focusable.
Start allowing this in case tabindex is set on form controls and adjust the tests accordingly.

* LayoutTests/fast/events/focus-label-legend-elements-with-tab-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-not-keyboard-focusable-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-not-keyboard-focusable.html:
* LayoutTests/platform/glib/fast/events/focus-label-legend-elements-with-tab-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/semantics/popovers/popover-not-keyboard-focusable-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-expected.txt:
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::isKeyboardFocusable const):
(WebCore::HTMLFormControlElement::isMouseFocusable const):

Canonical link: https://commits.webkit.org/263527@main
